packages/broker/test/unit/plugins/storage/Storage.test.ts

```diff
@@ -0,0 +1,255 @@
+import { Readable, PassThrough } from 'stream'
+import { Protocol } from 'streamr-network'
+import { Storage } from '../../../../src/plugins/storage/Storage'
+const { StreamMessage, MessageID } = Protocol.MessageLayer
+
+const MOCK_STREAM_ID = 'mock-stream-id'
+const MOCK_BUCKET = {
+    id: 'mock-bucket-id'
+}
+const MOCK_MESSAGE_COUNT = {
+    total: {
+        low: 1
+    }
+}
+const createMockMessage = (contentValue: number) => {
+    return {
+        payload: new StreamMessage({
+            messageId: new MessageID(MOCK_STREAM_ID, 0, Date.now(), 0, 'publisherId', 'msgChainId'),
+            content: {
+                value: contentValue
+            }
+        }).serialize()
+    }
+}
+const MOCK_MESSAGE_1 = createMockMessage(1)
+const MOCK_MESSAGE_2 = createMockMessage(2)
+const REQUEST_TYPE_FROM = 'requestFrom'
+const REQUEST_TYPE_RANGE = 'requestRange'
+
+const createResultFactory = ({
+    buckets,
+    messageCounts,
+    messages
+}: {
+    buckets?: { id: string}[]|Error,
+    messageCounts?: { total: { low: number } }[]|Error,
+    messages?: { payload: string }[]|Error
+}) => {
+    return (query: string) => {
+        if (query.includes('FROM bucket')) {
+            return buckets!
+        } else if (query.includes('total FROM stream_data')) {
+            return messageCounts!
+        } else if (query.includes('payload FROM stream_data')) {
+            return messages!
+        } else {
+            throw new Error(`Assertion failed: query="${query}"`)
+        }
+    }
+}
+
+const createMockStorage = (getResults: (query: string) => any[]|Error) => {
+    const cassandraClient = {
```

Comment:
This may be more in the lines of theoretical rambling, so take it as it as you will. 😄 

Given that you're mocking here, there could be a slight benefit of trying to have this variable be assignable (compatible with?) to `CassandraClient` interface / class as exposed by the library (as opposed to using `as any`). I.e.

```ts
let foo: cassandra.CassandraClientOrWhateverItIsCalled = cassandraClient // compiles
```

Let's say we decide to update the Cassandra driver and it contains a breaking change to `CassandraClient` that we rely on in our code. This test will still be green since we are testing against a mocked version of `CassandraClient`. However, if 
- (a) the implementation here would be assignable to `CassandraClient`
- (b) the breaking change came in the form of a breaking _type signature_ change

then there _could be a chance_ that the code here would not compile because our mocked version is not compatbile with the new `CassandraClient`.

This is of course quite theoretical, and not sure you can even do this in TypeScript. If it were doable without having to write boilerplate implementations of methods, i.e., the code snippet I pasted above "just worked" then could be interesting to try out 🤔

Reply:
The mock also has different async behaviour to the real Cassandra client.

 e.g.
 * `eachRow` here executes synchronously while it certainly doesn't with the real client, 
  * same with the `stream` implementation, the entire stream is created, filled, and ended synchronously
  * and maybe less of an issue but the mock `execute` promise will resolve in the next microtask, between `nextTick` and `setImmediate`, while in reality it's likely to always be some milliseconds later (after both `nextTick` and `setImmediate`).

Might not be an issue but having async behaviour that differs from the real implementation seems like a good way to hide (or even create) subtle async bugs during development.

IMO this should use a real Cassandra client unless there's some critical reason it can't.

Reply:
Need to have a client where it is easy to trigger all kinds of errors (promise rejections, `Readables` that emit error events). Easier with a mock instance compared to the real client. Agree that now we mock so much of Cassandra client that we don't really test the Cassandra interaction. But the tests validate our own query logic ok.

Reply:
Based on @timoxley's suggestions, refactored the test to use the real Cassandra client for non-error queries. Now it covers more Cassandra logic (and caught a bug https://github.com/streamr-dev/network-monorepo/pull/62). 

It would have been very convenient to implement the real vs. mock Cassandra client using jest `spy`. But as we needed to filter the success/error case by the query types (the parameter of a called method), it was more straightforward to implement the logic by a separate `ProxyClient` class. The class routes the query to the real client, or throws an `Error` if the test setup has called `setError(queryId)` for that query.